File: 2023/ruby/p09.rb
input = IO.foreach('../data/09.txt').map{|line| 
  line.split.map(&:to_i)
}

def f(seq)
  s = seq[-1]
  until seq.all?{_1 == 0}
    seq = seq[1..-1].zip(seq[0..-2]).map{|x, y| x - y}
    s += seq[-1]
  end
  s
end

# part 1
p input.map{f(_1)}.sum

# part 2
p input.map{f(_1.reverse)}.sum

File: 2023/nim/p09.nim
import strutils, sequtils, sugar, math, algorithm

let input = readFile("../data/09.txt").strip.splitLines.map(line =>
    line.splitWhitespace.mapIt(it.parseInt)
)

func f(xs: seq[int]): int =
    var s = xs[^1]
    var ys = xs
    while not ys.allIt(it == 0):
        ys = ys[1..^1].zip(ys[0..^2]).mapIt(it[0] - it[1])
        s += ys[^1]
    s

# part 1
echo input.map(f).sum

# part 2
echo input.mapIt(f(it.reversed)).sum
